credentials/tls/certprovider/meshca/builder.go

```diff
@@ -75,92 +75,83 @@ type pluginBuilder struct {
 	clients map[ccMapKey]*refCountedCC
 }
 
-// Build returns a MeshCA certificate provider for the passed in configuration
-// and options.
+// ParseConfig parses the configuration to be passed to the MeshCA plugin
+// implementation. Expects the config to be a json.RawMessage which contains a
+// serialized JSON representation of the meshca_experimental.GoogleMeshCaConfig
+// proto message.
 //
-// This builder takes care of sharing the ClientConn to the MeshCA server among
+// Takes care of sharing the ClientConn to the MeshCA server among
 // different plugin instantiations.
-func (b *pluginBuilder) Build(c certprovider.StableConfig, opts certprovider.Options) certprovider.Provider {
-	cfg, ok := c.(*pluginConfig)
+func (b *pluginBuilder) ParseConfig(c interface{}) (*certprovider.BuildableConfig, error) {
+	data, ok := c.(json.RawMessage)
 	if !ok {
-		// This is not expected when passing config returned by ParseConfig().
-		// This could indicate a bug in the certprovider.Store implementation or
-		// in cases where the user is directly using these APIs, could be a user
-		// error.
-		logger.Errorf("unsupported config type: %T", c)
-		return nil
+		return nil, fmt.Errorf("meshca: unsupported config type: %T", c)
 	}
-
-	b.mu.Lock()
-	defer b.mu.Unlock()
-
-	ccmk := ccMapKey{
-		name:    cfg.serverURI,
-		stsOpts: cfg.stsOpts,
+	cfg, err := pluginConfigFromJSON(data)
+	if err != nil {
+		return nil, err
 	}
-	rcc, ok := b.clients[ccmk]
-	if !ok {
-		// STS call credentials take care of exchanging a locally provisioned
-		// JWT token for an access token which will be accepted by the MeshCA.
-		callCreds, err := sts.NewCredentials(cfg.stsOpts)
-		if err != nil {
-			logger.Errorf("sts.NewCredentials() failed: %v", err)
-			return nil
-		}
+	return certprovider.NewBuildableConfig(pluginName, cfg.canonical(), func(opts certprovider.BuildOptions) certprovider.Provider {
```

Comment:
IMO this is too much code for a closure.

Can this be made a function instead, with parameters for whatever needs to be captured by the closure?

Reply:
Done.

Reply:
Ohh, I like how this makes the diffs much nicer as well.  😄